### ISSUE TITLE
for-of after an early return causes an NPE in `TypeInference`

### ISSUE THREAD

**Issue by kocoten1992** [google/closure-compiler#3844]
A nice calmed Sunday, I've been hacking on shaka-player and this error pop up, it tell me I should report this, here I am!

```
java.lang.RuntimeException: INTERNAL COMPILER ERROR.
Please report this problem.

null
  Node(FUNCTION ): /tmp/shaka-player/lib/text/vtt_text_parser.js:513:2
  static generateCueFromElement_(element, rootCue, cues, styles) {
  Parent(MEMBER_FUNCTION_DEF generateCueFromElement_): /tmp/shaka-player/lib/text/vtt_text_parser.js:513:9
  static generateCueFromElement_(element, rootCue, cues, styles) {

	at com.google.javascript.jscomp.TypeInference.updateScopeForAssignment(TypeInference.java:1163)
	at com.google.javascript.jscomp.TypeInference.updateScopeForAssignment(TypeInference.java:1139)
	at com.google.javascript.jscomp.TypeInference.branchedFlowThrough(TypeInference.java:520)
	at com.google.javascript.jscomp.TypeInference.branchedFlowThrough(TypeInference.java:84)
	at com.google.javascript.jscomp.DataFlowAnalysis$BranchedForwardDataFlowAnalysis.flow(DataFlowAnalysis.java:435)
	at com.google.javascript.jscomp.DataFlowAnalysis.analyze(DataFlowAnalysis.java:202)
	at com.google.javascript.jscomp.DataFlowAnalysis.analyze(DataFlowAnalysis.java:171)
	at com.google.javascript.jscomp.TypeInferencePass.inferScope(TypeInferencePass.java:142)
	at com.google.javascript.jscomp.TypeInferencePass.access$200(TypeInferencePass.java:32)
	at com.google.javascript.jscomp.TypeInferencePass$SecondScopeBuildingCallback.enterScope(TypeInferencePass.java:170)
	at com.google.javascript.jscomp.NodeTraversal.pushScope(NodeTraversal.java:1050)
	at com.google.javascript.jscomp.NodeTraversal.traverseFunction(NodeTraversal.java:917)
	at com.google.javascript.jscomp.NodeTraversal.handleFunction(NodeTraversal.java:814)
	at com.google.javascript.jscomp.NodeTraversal.traverseBranch(NodeTraversal.java:857)
	at com.google.javascript.jscomp.NodeTraversal.traverseChildren(NodeTraversal.java:1022)
	at com.google.javascript.jscomp.NodeTraversal.traverseBranch(NodeTraversal.java:890)
	at com.google.javascript.jscomp.NodeTraversal.handleClassMembers(NodeTraversal.java:1008)
	at com.google.javascript.jscomp.NodeTraversal.traverseBranch(NodeTraversal.java:866)
	at com.google.javascript.jscomp.NodeTraversal.handleClass(NodeTraversal.java:983)
	at com.google.javascript.jscomp.NodeTraversal.traverseBranch(NodeTraversal.java:863)
	at com.google.javascript.jscomp.NodeTraversal.traverseChildren(NodeTraversal.java:1022)
	at com.google.javascript.jscomp.NodeTraversal.traverseBranch(NodeTraversal.java:890)
	at com.google.javascript.jscomp.NodeTraversal.traverseChildren(NodeTraversal.java:1022)
	at com.google.javascript.jscomp.NodeTraversal.traverseBranch(NodeTraversal.java:890)
	at com.google.javascript.jscomp.NodeTraversal.traverseChildren(NodeTraversal.java:1022)
	at com.google.javascript.jscomp.NodeTraversal.handleScript(NodeTraversal.java:802)
	at com.google.javascript.jscomp.NodeTraversal.traverseBranch(NodeTraversal.java:854)
	at com.google.javascript.jscomp.NodeTraversal.traverseChildren(NodeTraversal.java:1022)
	at com.google.javascript.jscomp.NodeTraversal.traverseBranch(NodeTraversal.java:890)
	at com.google.javascript.jscomp.NodeTraversal.traverseChildren(NodeTraversal.java:1022)
	at com.google.javascript.jscomp.NodeTraversal.traverseBranch(NodeTraversal.java:890)
	at com.google.javascript.jscomp.NodeTraversal.traverseWithScope(NodeTraversal.java:526)
	at com.google.javascript.jscomp.NodeTraversal.access$600(NodeTraversal.java:38)
	at com.google.javascript.jscomp.NodeTraversal$Builder.traverseWithScope(NodeTraversal.java:414)
	at com.google.javascript.jscomp.TypeInferencePass.inferAllScopes(TypeInferencePass.java:116)
	at com.google.javascript.jscomp.DefaultPassConfig$6.process(DefaultPassConfig.java:1706)
	at com.google.javascript.jscomp.PhaseOptimizer$NamedPass.process(PhaseOptimizer.java:317)
	at com.google.javascript.jscomp.PhaseOptimizer.process(PhaseOptimizer.java:232)
	at com.google.javascript.jscomp.Compiler.check(Compiler.java:1030)
	at com.google.javascript.jscomp.Compiler.performChecksAndTranspilation(Compiler.java:828)
	at com.google.javascript.jscomp.Compiler.lambda$stage1Passes$0(Compiler.java:760)
	at com.google.javascript.jscomp.CompilerExecutor.lambda$runInCompilerThread$0(CompilerExecutor.java:101)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.NullPointerException
	... 46 more
[ERROR] Build failed
```

**Comment by brad4d** [comment]
@kocoten1992 it's nice of you to supply this stack trace, but I'm afraid we'd need more than just this to be able to investigate the problem.

What version of the compiler are you using?
Can you give us enough information so we can reproduce the problem?

If that's too much effort, feel free to close this issue and thanks anyway.

**Comment by kocoten1992** [comment]
@brad4d , I do have a concrete example (just not a minimal one - I'm sorry):
```bash
cd /tmp/
git clone https://github.com/google/shaka-player
git checkout 481b378a4bb249fb13c6a2f9916e38c859fcbe3f # it is the current master, I dont think it matter, just to be on the safe side
```
In `lib/text/vtt_text_parser.js`
```git
diff --git a/lib/text/vtt_text_parser.js b/lib/text/vtt_text_parser.js
index b9f36c38..7c98272e 100644
--- a/lib/text/vtt_text_parser.js
+++ b/lib/text/vtt_text_parser.js
@@ -537,6 +537,7 @@ shaka.text.VttTextParser = class {
         }
       }
     }
+    return;
     const isTextNode = shaka.util.XmlUtils.isText(element);
     if (isTextNode) {
       // Trailing line breaks may lost when convert cue to HTML tag
```

then run `python3 build/build.py --force`, it will bug out (without the edit - compile will success).

On my computer, the error look like this this:
```
[INFO] Compiling the library (ui, release)...
java.lang.RuntimeException: INTERNAL COMPILER ERROR.
Please report this problem.

null
  Node(FUNCTION ): /tmp/shaka-player/lib/text/vtt_text_parser.js:513:2
  static generateCueFromElement_(element, rootCue, cues, styles) {
  Parent(MEMBER_FUNCTION_DEF generateCueFromElement_): /tmp/shaka-player/lib/text/vtt_text_parser.js:513:9
  static generateCueFromElement_(element, rootCue, cues, styles) {

	at com.google.javascript.jscomp.TypeInference.updateScopeForAssignment(TypeInference.java:1163)
	at com.google.javascript.jscomp.TypeInference.updateScopeForAssignment(TypeInference.java:1139)
	at com.google.javascript.jscomp.TypeInference.branchedFlowThrough(TypeInference.java:520)
	at com.google.javascript.jscomp.TypeInference.branchedFlowThrough(TypeInference.java:84)
	at com.google.javascript.jscomp.DataFlowAnalysis$BranchedForwardDataFlowAnalysis.flow(DataFlowAnalysis.java:435)
	at com.google.javascript.jscomp.DataFlowAnalysis.analyze(DataFlowAnalysis.java:202)
	at com.google.javascript.jscomp.DataFlowAnalysis.analyze(DataFlowAnalysis.java:171)
	at com.google.javascript.jscomp.TypeInferencePass.inferScope(TypeInferencePass.java:142)
	at com.google.javascript.jscomp.TypeInferencePass.access$200(TypeInferencePass.java:32)
	at com.google.javascript.jscomp.TypeInferencePass$SecondScopeBuildingCallback.enterScope(TypeInferencePass.java:170)
	at com.google.javascript.jscomp.NodeTraversal.pushScope(NodeTraversal.java:1050)
	at com.google.javascript.jscomp.NodeTraversal.traverseFunction(NodeTraversal.java:917)
	at com.google.javascript.jscomp.NodeTraversal.handleFunction(NodeTraversal.java:814)
	at com.google.javascript.jscomp.NodeTraversal.traverseBranch(NodeTraversal.java:857)
	at com.google.javascript.jscomp.NodeTraversal.traverseChildren(NodeTraversal.java:1022)
	at com.google.javascript.jscomp.NodeTraversal.traverseBranch(NodeTraversal.java:890)
	at com.google.javascript.jscomp.NodeTraversal.handleClassMembers(NodeTraversal.java:1008)
	at com.google.javascript.jscomp.NodeTraversal.traverseBranch(NodeTraversal.java:866)
	at com.google.javascript.jscomp.NodeTraversal.handleClass(NodeTraversal.java:983)
	at com.google.javascript.jscomp.NodeTraversal.traverseBranch(NodeTraversal.java:863)
	at com.google.javascript.jscomp.NodeTraversal.traverseChildren(NodeTraversal.java:1022)
	at com.google.javascript.jscomp.NodeTraversal.traverseBranch(NodeTraversal.java:890)
	at com.google.javascript.jscomp.NodeTraversal.traverseChildren(NodeTraversal.java:1022)
	at com.google.javascript.jscomp.NodeTraversal.traverseBranch(NodeTraversal.java:890)
	at com.google.javascript.jscomp.NodeTraversal.traverseChildren(NodeTraversal.java:1022)
	at com.google.javascript.jscomp.NodeTraversal.handleScript(NodeTraversal.java:802)
	at com.google.javascript.jscomp.NodeTraversal.traverseBranch(NodeTraversal.java:854)
	at com.google.javascript.jscomp.NodeTraversal.traverseChildren(NodeTraversal.java:1022)
	at com.google.javascript.jscomp.NodeTraversal.traverseBranch(NodeTraversal.java:890)
	at com.google.javascript.jscomp.NodeTraversal.traverseChildren(NodeTraversal.java:1022)
	at com.google.javascript.jscomp.NodeTraversal.traverseBranch(NodeTraversal.java:890)
	at com.google.javascript.jscomp.NodeTraversal.traverseWithScope(NodeTraversal.java:526)
	at com.google.javascript.jscomp.NodeTraversal.access$600(NodeTraversal.java:38)
	at com.google.javascript.jscomp.NodeTraversal$Builder.traverseWithScope(NodeTraversal.java:414)
	at com.google.javascript.jscomp.TypeInferencePass.inferAllScopes(TypeInferencePass.java:116)
	at com.google.javascript.jscomp.DefaultPassConfig$6.process(DefaultPassConfig.java:1706)
	at com.google.javascript.jscomp.PhaseOptimizer$NamedPass.process(PhaseOptimizer.java:317)
	at com.google.javascript.jscomp.PhaseOptimizer.process(PhaseOptimizer.java:232)
	at com.google.javascript.jscomp.Compiler.check(Compiler.java:1030)
	at com.google.javascript.jscomp.Compiler.performChecksAndTranspilation(Compiler.java:828)
	at com.google.javascript.jscomp.Compiler.lambda$stage1Passes$0(Compiler.java:760)
	at com.google.javascript.jscomp.CompilerExecutor.lambda$runInCompilerThread$0(CompilerExecutor.java:101)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.NullPointerException
	... 46 more
[ERROR] Build failed

```

As for version, I think it is https://github.com/google/shaka-player/blob/master/package.json#L31-L32

**Comment by arnav-kr** [comment]
i got the same error when using the REST API

```
"java.lang.RuntimeException: INTERNAL COMPILER ERROR.
Please report this problem.

java.lang.InterruptedException
	at com.google.javascript.jscomp.NodeTraversal.initTraversal(NodeTraversal.java:1305)
	at com.google.javascript.jscomp.NodeTraversal.traverse(NodeTraversal.java:455)
	at com.google.javascript.jscomp.NodeTraversal.access$200(NodeTraversal.java:38)
	at com.google.javascript.jscomp.NodeTraversal$Builder.traverse(NodeTraversal.java:398)
	at com.google.javascript.jscomp.NodeTraversal.traverse(NodeTraversal.java:468)
	at com.google.javascript.jscomp.ControlFlowAnalysis.process(ControlFlowAnalysis.java:159)
	at com.google.javascript.jscomp.TypeInferencePass.computeCfg(TypeInferencePass.java:170)
	at com.google.javascript.jscomp.TypeInferencePass.inferScope(TypeInferencePass.java:124)
	at com.google.javascript.jscomp.TypeInferencePass.access$200(TypeInferencePass.java:31)
	at com.google.javascript.jscomp.TypeInferencePass$SecondScopeBuildingCallback.enterScope(TypeInferencePass.java:158)
	at com.google.javascript.jscomp.NodeTraversal.pushScope(NodeTraversal.java:1028)
	at com.google.javascript.jscomp.NodeTraversal.traverseFunction(NodeTraversal.java:895)
	at com.google.javascript.jscomp.NodeTraversal.handleFunction(NodeTraversal.java:792)
	at com.google.javascript.jscomp.NodeTraversal.traverseBranch(NodeTraversal.java:835)
	at com.google.javascript.jscomp.NodeTraversal.traverseChildren(NodeTraversal.java:1000)
	at com.google.javascript.jscomp.NodeTraversal.traverseBranch(NodeTraversal.java:868)
	at com.google.javascript.jscomp.NodeTraversal.traverseChildren(NodeTraversal.java:1000)
	at com.google.javascript.jscomp.NodeTraversal.traverseBranch(NodeTraversal.java:868)
	at com.google.javascript.jscomp.NodeTraversal.traverseChildren(NodeTraversal.java:1000)
	at com.google.javascript.jscomp.NodeTraversal.handleScript(NodeTraversal.java:780)
	at com.google.javascript.jscomp.NodeTraversal.traverseBranch(NodeTraversal.java:832)
	at com.google.javascript.jscomp.NodeTraversal.traverseChildren(NodeTraversal.java:1000)
	at com.google.javascript.jscomp.NodeTraversal.traverseBranch(NodeTraversal.java:868)
	at com.google.javascript.jscomp.NodeTraversal.traverseChildren(NodeTraversal.java:1000)
	at com.google.javascript.jscomp.NodeTraversal.traverseBranch(NodeTraversal.java:868)
	at com.google.javascript.jscomp.NodeTraversal.traverseWithScope(NodeTraversal.java:522)
	at com.google.javascript.jscomp.NodeTraversal.access$500(NodeTraversal.java:38)
	at com.google.javascript.jscomp.NodeTraversal$Builder.traverseWithScope(NodeTraversal.java:410)
	at com.google.javascript.jscomp.TypeInferencePass.inferAllScopes(TypeInferencePass.java:104)
	at com.google.javascript.jscomp.DefaultPassConfig.lambda$new$32(DefaultPassConfig.java:1631)
	at com.google.javascript.jscomp.PhaseOptimizer$NamedPass.process(PhaseOptimizer.java:317)
	at com.google.javascript.jscomp.PhaseOptimizer.process(PhaseOptimizer.java:232)
	at com.google.javascript.jscomp.Compiler.check(Compiler.java:1063)
	at com.google.javascript.jscomp.Compiler.performChecksAndTranspilation(Compiler.java:861)
	at com.google.javascript.jscomp.Compiler.lambda$stage1Passes$4(Compiler.java:793)
	at com.google.javascript.jscomp.CompilerExecutor.runInCompilerThread(CompilerExecutor.java:127)
	at com.google.javascript.jscomp.Compiler.runInCompilerThread(Compiler.java:849)
	at com.google.javascript.jscomp.Compiler.stage1Passes(Compiler.java:791)
	at com.google.javascript.jscomp.Compiler.compile(Compiler.java:708)
	at com.google.javascript.jscomp.webservice.backend.CompilerInvokerImpl.compile(CompilerInvokerImpl.java:44)
	at com.google.javascript.jscomp.webservice.backend.ServerController.executeRequest(ServerController.java:178)
	at com.google.javascript.jscomp.webservice.backend.CompilationRequestHandler.serviceParsedRequest(CompilationRequestHandler.java:178)
	at com.google.javascript.jscomp.webservice.backend.CompilationRequestHandler.service(CompilationRequestHandler.java:160)
	at com.google.javascript.jscomp.webservice.frontend.CompilationServlet.doPost(CompilationServlet.java:82)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:707)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:848)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1780)
	at com.google.apphosting.utils.servlet.JdbcMySqlConnectionCleanupFilter.doFilter(JdbcMySqlConnectionCleanupFilter.java:62)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1767)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:583)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:524)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:226)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:134)
	at com.google.apphosting.runtime.jetty9.ParseBlobUploadHandler.handle(ParseBlobUploadHandler.java:111)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1182)
	at com.google.apphosting.runtime.jetty9.AppEngineWebAppContext.doHandle(AppEngineWebAppContext.java:187)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:513)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1112)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at com.google.apphosting.runtime.jetty9.AppVersionHandlerMap.handle(AppVersionHandlerMap.java:293)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:134)
	at org.eclipse.jetty.server.Server.handle(Server.java:539)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:333)
	at com.google.apphosting.runtime.jetty9.RpcConnection.handle(RpcConnection.java:216)
	at com.google.apphosting.runtime.jetty9.RpcConnector.serviceRequest(RpcConnector.java:81)
	at com.google.apphosting.runtime.jetty9.JettyServletEngineAdapter.serviceRequest(JettyServletEngineAdapter.java:134)
	at com.google.apphosting.runtime.JavaRuntime$RequestRunnable.dispatchServletRequest(JavaRuntime.java:794)
	at com.google.apphosting.runtime.JavaRuntime$RequestRunnable.dispatchRequest(JavaRuntime.java:757)
	at com.google.apphosting.runtime.JavaRuntime$RequestRunnable.run(JavaRuntime.java:727)
	at com.google.apphosting.runtime.ThreadGroupPool$PoolEntry.run(ThreadGroupPool.java:261)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.RuntimeException: java.lang.InterruptedException
	... 75 more
Caused by: java.lang.InterruptedException
	... 75 more
"```
